src/sinks/greptimedb/logs/sink.rs

```diff
@@ -0,0 +1,88 @@
+use crate::sinks::greptimedb::logs::http_request_builder::{
+    GreptimeDBLogsHttpRequestBuilder, KeyPartitioner, PartitionKey,
+};
+use crate::sinks::prelude::*;
+use crate::sinks::util::http::HttpRequest;
+
+/// A sink that ingests logs into GreptimeDB.
+pub struct GreptimeDBLogsHttpSink<S> {
+    batcher_settings: BatcherSettings,
+    service: S,
+    dbname: Template,
+    table: Template,
+    pipeline_name: Template,
+    pipeline_version: Option<Template>,
+    request_builder: GreptimeDBLogsHttpRequestBuilder,
+}
+
+impl<S> GreptimeDBLogsHttpSink<S>
+where
+    S: Service<HttpRequest<PartitionKey>> + Send + 'static,
+    S::Future: Send + 'static,
+    S::Response: DriverResponse + Send + 'static,
+    S::Error: std::fmt::Debug + Into<crate::Error> + Send,
+{
+    pub const fn new(
+        batcher_settings: BatcherSettings,
+        service: S,
+        dbname: Template,
+        table: Template,
+        pipeline_name: Template,
+        pipeline_version: Option<Template>,
+        request_builder: GreptimeDBLogsHttpRequestBuilder,
+    ) -> Self {
+        Self {
+            batcher_settings,
+            service,
+            dbname,
+            table,
+            pipeline_name,
+            pipeline_version,
+            request_builder,
+        }
+    }
+
+    async fn run_inner(self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        let batcher_settings = self.batcher_settings;
+        input
+            .batched_partitioned(
+                KeyPartitioner::new(
+                    self.dbname,
+                    self.table,
+                    self.pipeline_name,
+                    self.pipeline_version,
+                ),
+                || batcher_settings.as_byte_size_config(),
+            )
+            .filter_map(|(key, batch)| async move { key.map(move |k| (k, batch)) })
+            .request_builder(
+                default_request_builder_concurrency_limit(),
+                self.request_builder,
+            )
+            .filter_map(|request| async {
+                match request {
+                    Err(error) => {
+                        emit!(SinkRequestBuildError { error });
+                        None
+                    }
+                    Ok(req) => Some(req),
+                }
+            })
+            .into_driver(self.service)
```

Comment:
Can we add a call to `.protocol` here. This is used by the driver to emit `bytes_sent` metrics. I believe this should either be `http` or `https` depending on the scheme in the endpoint. You can have a look at the [`datodog_logs`](https://github.com/vectordotdev/vector/blob/master/src/sinks/datadog/logs/sink.rs#L370) sink as a reference.

Reply:
I realized that I can't call protocol here. when i run test.  I got this error. This function `run_and_assert_sink_compliance` seems to check for something.

```
  - Multiple (2) events matching `BytesSent`: (`BytesSent`, `EndpointBytesSent`). Hint! Don't use the `assert_x_` test helpers on round-trip tests (tests that run more than a single component).
```

`HttpBatchService` adds a events `EndpointBytesSent` to `EVENTS_RECORDED` here.
https://github.com/vectordotdev/vector/blob/41bb21ef711d55884b02eee42b0126c25e97dd5e/src/sinks/util/http.rs#L422
If I call protocol here, the following code adds `BytesSent` to `EVENTS_RECORDED`. This causes the error in the test above.
https://github.com/vectordotdev/vector/blob/abd776d7c74ae48968fa34829d3683f68115a9e0/lib/vector-stream/src/driver.rs#L99